### ISSUE TITLE
suggestion: unix-common platform

### ISSUE THREAD

**Issue by alexbuzzbee** [tldr-pages/tldr#1868]
We currently have a `common` platform that contains a number of commands not available on Windows. A number of commands (e.g. `tar`) are available on all Unixes, but not Windows. As such, I suggest we create a `unix-common` platform to contain commands common to Unix platforms. `unix-common` would be checked/cached by clients running on Unix platforms, but not by clients on Windows. Once this is working, all the Unix commands present either in `common` or duplicated across the `linux`, `osx`, and `sunos` would be moved to `unix-common` and only commands available on Unix and Windows would be left in `common`.

**Comment by agnivade** [comment]
Good idea. But I fear this will need a lot of manual work of checking which pages are common to all and which are common to unix only. And also, _all_ of the clients need to update their code.

But it's a start nonetheless. Duplicating pages across folders is just plain ugly. 

@waldyrious / @sbrl - what do you think ?

**Comment by sbrl** [comment]
Hey! This must have slipped through the net. Yeah, I'd be all for a `unix-common`, as there are a considerable number of commands that are common to all unix-like POSIX systems that aren't the same on Windows.

**Comment by zdroid** [comment]
Can we merge this issue with #1966? That issue already includes this problem.

**Comment by sbrl** [comment]
I think it might be worth avoiding merging @zdroid, just to keep things a little less confusing. If you were to merge, you could technically even include that other issue with recent activity about a flatter structure. All issues are related to each other in some fashion - even if not directly :P

I'd suggest that in this case it's helpful to have a separate issue about a specific aspect of it, since it's easier to keep track that way.

**Comment by zdroid** [comment]
All issues are connected, that's true, but this and #1966 are really closely connected. This issue's proposal is a solution for a huge part of #1966 we seem to agree on (at least a bit).

**Comment by sbrl** [comment]
We should also consider how we'd need to update the client spec for this. The client spec outlines the canonical page resolution algorithm (the algorithm that determines which page a client should show) - which would need to be updated.

**Comment by zdroid** [comment]
Would it actually? Technically, all pages could stay in `common/` as they are now.

The only reason for algorithm to change is if someone wants to add a specific Unix flag.

**Comment by alexbuzzbee** [comment]
Happy second birthday to this issue. 🎂

**Comment by alexbuzzbee** [comment]
One possibility I've thought of would be to have a `multi` (`shared`?) directory with the originals of commands from multiple platforms, then symlink from the platform directories to the `multi` pages. That would eliminate the duplication and prevent them from getting out of sync, and it wouldn't require such strict control of what goes in as `common`/`unix-common`.

**Comment by sbrl** [comment]
It would @zdroid, since Linux is a UNIX-based platform for example - and under the current tldr-pages page resolution algorithm it would not show it - since it would only look at the platforms `linux` and then `common`.

@alexbuzzbee: That's sort of the idea of the `common` platform. I believe we've discussed symlinks before though - and I'm in favour of them - but apparently some platforms do not support symlinks, so it's a bit of a challenge to use them in the context of tldr-pages.

**Comment by cyqsimon** [comment]
Hi all I'd like to revive this issue and kickstart it with a PR.

Currently there are just under 5000 English pages, which is a lot of work but still manageable. The more we delay this the more work this will entail.

**Comment by waldyrious** [comment]
@sebastiaanspeck I'm sorry to interject so sporadically in the project, but I couldn't help but notice that you closed this issue without any comment. Was there discussion elsewhere that led to the closure? If so, please link to it for future reference.

As for @cyqsimon's proposal to revive the discussion — thanks for the initiative! I'll add a suggestion that I don't recall seeing before: how about we use a tagging system rather than a filesystem hierarchy one? That way we wouldn't have to create all the possible directory combinations for e.g. commands that exist on linux and macos, but not *bsd, or exist for macos and windows but not linux, and so on (or worse, duplicate pages if they fit more than one existing category).

Of course, doing so would require a large restructuring of the project, with a flat hierarchy where all pages would be placed in the same top-level directory, or perhaps grouped into alphabetically sorted prefix directories, as projects like [repology](https://github.com/repology/repology-rules/tree/master/850.split-ambiguities) do. So it's not a decision to be taken lightly, but it's food for thought as a possible approach to tackle this problem.

**Comment by waldyrious** [comment]
I'd say that reinforces my point precisely: documenting why decisions are made allows us to regain context even months or years in the future (not to mention being more respectful of the participants :slightly_smiling_face:). I would recommend reopening this issue unless/until the actual reason for closure is explicitly mentioned, and/or decision-making trail is recovered.

**Comment by kbdharun** [comment]
> I'd say that reinforces my point precisely: documenting why decisions are made allows us to regain context even months or years in the future (not to mention being more respectful of the participants 🙂). I would recommend reopening this issue unless/until the actual reason for closure is explicitly mentioned, and/or decision-making trail is recovered.

Agreed, I do remember us having a conversation in a PR/issue or in the chatroom regarding this and this was closed after it (we might have forgotten to comment about the same here).

---

The current "common" directory now targets the POSIX variants of commands making the UNIX-common platform addition redundant since it will be a subset of the existing common platform, we are documenting platform-specific commands under separate dedicated pages under the platform.

> We currently have a common platform that contains a number of commands not available on Windows. A number of commands (e.g. tar) are available on all Unixes, but not Windows. As such, I suggest we create a unix-common platform to contain commands common to Unix platforms. unix-common would be checked/cached by clients running on Unix platforms, but not by clients on Windows. 

Regarding the initial proposal, one other thing I remember coming up in the previous conversation was since WSL's introduction in Windows, many UNIX commands were increasingly being used by our users so this change would lead to fragmentation with them (since there are issues like https://github.com/dbrgn/tealdeer/issues/377).

**Comment by sbrl** [comment]
Just to add my thoughts here. It is generally my experience that commands we document are a lot messier than just "UNIX / POSIX" and "Windows" / "Something else". For example, Image Magick (`magick` etc) is available for many platforms: <https://imagemagick.org/script/download.php>

...hence, adding `unix-common` would be a complicated process that is rather subjective - especially with tools like cygwin, PowerShell, msys, and others that bring traditionally Linux commands to Windows and others.

This is in contrast to my earlier opinion I posted above in 2018.
